Sources/Commands/SwiftTool.swift

```diff
@@ -452,42 +452,63 @@ public class SwiftTool {
+        // 👀 is multiroot-data-file in use?
         if let multiRootPackageDataFile = options.multirootPackageDataFile {
             return multiRootPackageDataFile.appending(components: "xcshareddata", "swiftpm", "Package.resolved")
         }
-        return try getPackageRoot().appending(component: "Package.resolved")
+        return try Workspace.DefaultLocations.resolvedVersionsFile(forRootPackage: self.getPackageRoot())
     }
 
-    func mirrorsConfigFilePath() throws -> AbsolutePath {
+    func getMirrorsConfig(sharedConfigurationDirectory: AbsolutePath? = nil) throws -> Configurations.Configuration.WorkspaceMirrors {
+        let sharedConfigurationDirectory = try sharedConfigurationDirectory ?? self.getSharedConfigurationDirectory()
+        let sharedMirrorFile = sharedConfigurationDirectory.map { Workspace.DefaultLocations.mirrorsConfigurationFile(at: $0) }
+        return try .init(
+            localMirrorFile: self.mirrorsConfigFile(),
+            sharedMirrorFile: sharedMirrorFile,
+            fileSystem: localFileSystem
+        )
+    }
+
+    private func mirrorsConfigFile() throws -> AbsolutePath {
+        // 👀 does this make sense now that we a global configuration as well? or should we at least rename it?
         // Look for the override in the environment.
         if let envPath = ProcessEnv.vars["SWIFTPM_MIRROR_CONFIG"] {
             return try AbsolutePath(validating: envPath)
         }
 
         // Otherwise, use the default path.
+        // 👀 is multiroot-data-file in use?
         if let multiRootPackageDataFile = options.multirootPackageDataFile {
-            return multiRootPackageDataFile.appending(components: "xcshareddata", "swiftpm", "config")
+            // migrate from legacy location
+            let legacyPath = multiRootPackageDataFile.appending(components: "xcshareddata", "swiftpm", "config")
+            let newPath = multiRootPackageDataFile.appending(components: "xcshareddata", "swiftpm", "configuration", "mirrors.json")
+            if localFileSystem.exists(legacyPath) {
+                try localFileSystem.createDirectory(newPath.parentDirectory, recursive: true)
+                try localFileSystem.move(from: legacyPath, to: newPath)
+            }
+            return newPath
         }
-        return try getPackageRoot().appending(components: ".swiftpm", "config")
-    }
 
-    func getMirrorsConfig() throws -> Workspace.Configuration {
-        return try _mirrorsConfig.get()
+        // migrate from legacy location
+        let legacyPath = try self.getPackageRoot().appending(components: ".swiftpm", "config")
+        let newPath = try Workspace.DefaultLocations.mirrorsConfigurationFile(forRootPackage: self.getPackageRoot())
+        if localFileSystem.exists(legacyPath) {
+            try localFileSystem.createDirectory(newPath.parentDirectory, recursive: true)
+            try localFileSystem.move(from: legacyPath, to: newPath)
+        }
+        return newPath
```

Comment:
@mattt @neonichu @abertelrud note the code above does the migration from old path to new one. we should carefully integrate that into the clients such as Xcode

Reply:
Among the other development-oriented documents in SwiftPM, maybe we should have one specifically focused on clients like IDEs that calls out these things.  Or maybe a separate section in the change log.  Just a thought.